### PR TITLE
fix(router): encode URLs the same way AngularJS did (closer to spec)

### PR DESCRIPTION
**Pull request by jasonaden**

### CODE REVIEW

File: packages/router/src/url_tree.ts [hunk 1]
```diff
@@ -325,8 +325,24 @@ function serializeSegment(segment: UrlSegmentGroup, root: boolean): string {
...
+ *    pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
+ *    unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
+ *    pct-encoded   = "%" HEXDIG HEXDIG
+ *    sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
+ *                     / "*" / "+" / "," / ";" / "="
+ */
 export function encode(s: string): string {
-  return encodeURIComponent(s);
+  return encodeURIComponent(s)
+      .replace(/%40/gi, '@')
```
Comment (vicb) [inline packages/router/src/url_tree.ts]: extra i

File: packages/router/test/url_serializer.spec.ts [hunk 2]
```diff
@@ -219,6 +219,17 @@ describe('url serializer', () => {
       expect(url.serialize(tree)).toEqual(u);
     });
 
+    it('should encode query params leaving sub-delimiters intact', () => {
+      const percentChars = '/?#[]&+= ';
+      const percentCharsConverted = '%2F%3F%23%5B%5D%26%2B%3D%20';
+      const intactChars = '!$\'()*,;:';
+      const params = percentChars + intactChars;
+
+      expect(percentCharsConverted).toEqual(encode(percentChars));
```
Comment (vicb) [inline packages/router/test/url_serializer.spec.ts]: case

File: packages/router/test/url_serializer.spec.ts [hunk 3]
```diff
@@ -219,6 +219,17 @@ describe('url serializer', () => {
...
+      expect(intactChars).toEqual(encode(intactChars));
+      expect(percentCharsConverted + intactChars).toEqual(encode(params));
```
Comment (vicb) [inline packages/router/test/url_serializer.spec.ts]: why ?

File: packages/router/src/url_tree.ts [hunk 4]
```diff
@@ -325,8 +325,25 @@ function serializeSegment(segment: UrlSegmentGroup, root: boolean): string {
...
+ *    pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
+ *    unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
+ *    pct-encoded   = "%" HEXDIG HEXDIG
+ *    sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
+ *                     / "*" / "+" / "," / ";" / "="
+ */
 export function encode(s: string): string {
-  return encodeURIComponent(s);
+  return encodeURIComponent(s)
+      .toUpperCase()
```
Comment (vicb) [inline packages/router/src/url_tree.ts]: you don't want `.toUpperCase()` outside of encoded char (`low` should remain `low` -> add test)

File: packages/router/test/url_serializer.spec.ts [hunk 5]
```diff
@@ -219,6 +219,19 @@ describe('url serializer', () => {
       expect(url.serialize(tree)).toEqual(u);
     });
 
+    it('should encode query params leaving sub-delimiters intact', () => {
+      const percentChars = '/?#[]&+= ';
+      const percentCharsConverted = '%2F%3F%23%5B%5D%26%2B%3D%20';
```
Comment (vicb) [inline packages/router/test/url_serializer.spec.ts]: Converted -> Encoded ?

File: packages/router/src/url_tree.ts (hunk 4 above)
Comment (jasonaden) [inline packages/router/src/url_tree.ts]: Yeah, for sure. That broke a test or two. I'll add an expectation.

File: packages/router/test/url_serializer.spec.ts (hunk 5 above)
Comment (jasonaden) [inline packages/router/test/url_serializer.spec.ts]: Fixed

**Review by vicb — APPROVED** [review]

**Pull request merged**